include/drivers/rtio_sensor.h

```diff
@@ -0,0 +1,127 @@
+ * values.
+ *
+ * Each RTIO Sensor is expected to output buffers containing sample sets (such
+ * as 9 axis IMU data) in a device specific format understood by the driver.
+ * Each driver must then implement the sensor reader api and functionality
+ * to deserialize this stream of buffers into meaningful SI unit data. The
+ * caller may request different numerical formats of this data.
+ *
+ * *TODO* To start only 32bit floating point numerical formatted values are
+ * provided. Q format values with scale factors would be very useful for
+ * use with CMSIS DSP or many SoCs that do not contain FPUs. The reader
+ * *should* provide simple ways of converting integer readings into
+ * useful numerical formats in common functions.
+ */
+
+/**
+ * @private
+ */
+struct rtio_sensor_reader;
+
+/**
+ * @private
+ * @brief Get the number of sample sets in the current block
+ */
+typedef ssize_t (*rtio_sensor_reader_sample_sets_t)(const struct sensor_reader *reader);
+
+/**
+ * @private
+ * @brief Iterate to the next sample set in the block
+ *
+ * Returns 0 when complete, 1 otherwise
+ *
+ */
+typedef int (*rtio_sensor_reader_next_t)(struct rtio_sensor_reader *reader);
+
+/**
+ * @private
+ * @brief Get the value of a sample sets channel, if it exists
+ *
+ * The channel is an index into a device specified sample channel.
+ * For example, A 3 axis accelerometer may assign channel 0 to
+ * the X axis, 1 to the Y axis, and 2 to the Z axis.
+ *
+ * *TODO* discuss how the channel identifier might be obtained for something like
+ * the X channel so that the same API calls may be used for multiple devices.
+ */
+typedef int (*rtio_sensor_reader_get_float_t)(const struct sensor_reader *reader,
+					      const u32_t channel, float *val);
+
+
+/**
+ * @private
+ * @brief Calculate or return the timestamp of the current sample set
+ *
+ * The timestamp may be part of the sample set or may be interpolated from the
+ * blocks begin and end timestamps along with the number of sample sets in the
+ * block or it may be in the block data itself.
+ */
+typedef int (*rtio_sensor_reader_timestamp_t)(const struct sensor_reader *reader,
+					      u32_t *timestamp);
+
+/**
+ * @brief RTIO Sensor Reader
+ */
+struct rtio_sensor_reader {
+	rtio_sensor_sample_sets_t sample_sets;
+	rtio_sensor_reader_next_t next;
+	rtio_sensor_reader_get_float_t get_float;
+	rtio_sensor_reader_timestamp_t timestamp;
+
+	struct rtio_block *block;
+	size_t position;
+};
+
+/**
+ * @private
+ * @brief Initialize a block reader with the given device and block
+ *
+ * The caller is still responsible for freeing the rtio_block.
+ */
+typedef int (*rtio_sensor_reader_t)(struct device *dev, struct rtio_block *block, struct sensor_reader *reader, u32_t *timestamp);
+
+/**
+ * @brief Real-Time IO Sensor API
+ */
+struct rtio_sensor_api {
+	/** Configuration function pointer *must* be implemented */
+	rtio_api rtio;
+
+	/** RTIO Block Reader */
+	rtio_sensor_reader_t block_reader;
```

Comment:
Can you explain why you did it this way? I was thinking that the RTIO sensor API would be an extension of the RTIO API, kind of like a subclass or Peter's idea of providing device-specific functionality via embedding the common API struct within a larger one.

Reply:
Thats exactly what this is

Reply:
Is it? It looks to me like `struct rtio_sensor_reader` is a combination of an API struct and some RTIO sensor device-specific state. I would have expected something like this:

```
struct rtio_sensor_api {
	struct rtio_api;
	rtio_sensor_sample_sets_t sample_sets;
	rtio_sensor_reader_next_t next;
	rtio_sensor_reader_get_float_t get_float;
	rtio_sensor_reader_timestamp_t timestamp;
};
```

And then a `struct device *dev` that was an RTIO sensor would have `dev->driver_api` point to a `struct rtio_sensor_api`. That way you could treat it as an ordinary RTIO device with additional features.

Is that not how those extensions were meant to work?

Reply:
So the above API assumes the driver will create the same reader for every block. Blocks can have unique layouts that may vary the reader returned. So for each block recieved it made sense, to me at least, to allow the driver to vary the reader at that time depending on the block layout.